test/integration/framework/etcd.go

```diff
@@ -182,37 +182,22 @@ func EtcdMain(tests func() int) {
 	// Bail out early when -help was given as parameter.
 	flag.Parse()
 
-	before := runtime.NumGoroutine()
+	// Ignore all goroutines created by "go test" and init functions,
+	// like the one from go.opencensus.io/stats/view/worker.go.
+	goleakOpts := []goleak.Option{goleak.IgnoreCurrent()}
+
 	stop, err := startEtcd()
 	if err != nil {
 		klog.Fatalf("cannot run integration tests: unable to start etcd: %v", err)
 	}
 	result := tests()
 	stop() // Don't defer this. See os.Exit documentation.
 
-	checkNumberOfGoroutines := func() (bool, error) {
-		// We leave some room for leaked goroutines as there are
-		// still some leaks, mostly:
-		// - leak from lumberjack package we're vendoring
-		// - leak from apiserve healthz
-		// - leak from opencensus library
-		// Once fixed, we should be able to bring it down to zero.
-		if dg := runtime.NumGoroutine() - before; dg <= 3 {
-			return true, nil
-		}
-		// Allow goroutines to schedule and die off.
-		runtime.Gosched()
-		return false, nil
+	if err := goleak.Find(goleakOpts...); err != nil {
+		klog.ErrorS(err, "EtcdMain goroutine check")
+		result = 1
 	}
 
-	// It generally takes visibly less than 1s to finish all goroutines.
-	// But we keep the limit higher to account for cpu-starved environments.
-	if err := wait.Poll(100*time.Millisecond, 5*time.Second, checkNumberOfGoroutines); err != nil {
-		after := runtime.NumGoroutine()
-		stacktraces := make([]byte, 1<<20)
-		runtime.Stack(stacktraces, true)
-		klog.Fatalf("unexpected number of goroutines: before: %d after %d\n%sd", before, after, string(stacktraces))
```

Comment:
Note that replacing this code also fixes a slight bug: the string that gets written here is `1<<20` bytes long.

The code should have used:
```
len := runtime.Stack(stacktraces, true)
... string(stacktraces[:len]) ...
```

Reply:
you are awesome 👏